### ISSUE TITLE
Use PyTorch data loading infrastructure?

### ISSUE THREAD

**Issue by aleSuglia** [allenai/allennlp#5090]
Hello,

I love AllenNLP and I incredibly appreciate the work that AllenNLP group at AI2 has done on this :)

However, I feel like recently I've been constantly dealing with bugs due to weird behaviours in the data loading process. So I started wondering: is there any specific reason why we had to reimplement the entire data loading logic? I was thinking to a very simple adaptation of the AllenNLP infrastructure. This is my personal opinion and might be heavily biased by my use cases therefore please let me know if I'm wrong.

Currently, we implement a `DatasetReader` that deals with transforming text to `Instance` objects. Most of the time, these instances are incredibly big and occupy a lot of memory making training on large datasets pretty much impossible. My experience with lazy loading hasn't been very successful so far with workers hanging waiting for others to actually generate the instances.

So I was thinking to the following solution that can be summarised as follows:

1. Implement a PyTorch `Dataset` or `IterableDataset` that handles the logic of reading the raw data. The user will decide whether they want a lazy dataset or not. I believe that this logic can be somehow abstracted away later on. For simplicity let's assume that the user decides this and not AllenNLP;
2. Implement a `DatasetTransform` that has a method `text_to_instance` which, given a raw input datum, returns an `Instance` object;
3. The `__get_item__` of a `Dataset` calls in turn `text_to_instance` of the `DatasetTransform` specifically designed for that dataset. A `Dataset` will assume that the raw data have been loaded in memory as a list (or dictionary). An `IterableDataset` can be easily be implemented following the default `PyTorch` logic;
4. The Trainer uses a PyTorch `DataLoader` to load the instances and uses `allennlp_collate_fn` to batch the data following the AllenNLP padding strategy.

For all these component we can create AllenNLP registrable wrappers so that they can be easily integrated in an AllenNLP experiment. In this way, we can literally reuse the already robust and reliable backbone infrastructure that PyTorch offers still benefiting from everything that you guys have already beautifully implemented. 

@epwalsh @dirkgr @matt-gardner Am I oversimplifying this? I'd love to hear your thoughts on this!

Thanks a lot for your help,
Alessandro

**Comment by epwalsh** [comment]
@aleSuglia I'll try my best to answer your concerns and questions, and explain our decision making process behind the current data loading API, which is certainly not perfect, but I do believe is still a significant improvement over the old system.

> recently I've been constantly dealing with bugs due to weird behaviours in the data loading process

AllenNLP's `MultiProcessDataLoader` uses the exact same tools under the hood as the PyTorch data loader, but is actually a lot simpler. We only use a single queue, whereas PyTorch has multiple queues. See https://github.com/pytorch/pytorch/blob/3e185253b65be70162e4bcc8778c7972d5ddf86d/torch/utils/data/dataloader.py#L571-L585, for example.

So any hangs or dead-locking issues that could happen with AllenNLP's `MultiProcessDataLoader` are at least as likely to occur with PyTorch's data loader. See https://github.com/allenai/allennlp/discussions/4935 for solutions to some common issues.

> is there any specific reason why we had to reimplement the entire data loading logic?

Yes, a couple of reasons:

1. The PyTorch data loader does not support batch samplers with an iterable dataset.
2. Multiprocess data loading with the PyTorch data loader in AllenNLP did not work how you'd expect with regular datasets. No matter how many workers you used, only a single process was responsible for reading the data into memory. So there was really no speed up from using more than one worker.

But there is nothing stopping you from utilizing the PyTorch data loader within AllenNLP. AllenNLP's `DataLoader` is a registrable class, so you could simply implement your own subclass that wraps the PyTorch data loader. We did actually have this for a while, but took it out because we didn't see any benefit of it.

> these instances are incredibly big and occupy a lot of memory making training on large datasets pretty much impossible

Why are your instances so big? Have you tried just using `TensorFields` in your instances instead of the higher level fields like `TextField` and `ListField`?

`Instance` objects themselves are light-weight, it's what you put into them that determines their size.

> My experience with lazy loading hasn't been very successful so far with workers hanging waiting for others to actually generate the instances.

What workers are waiting? If you're doing lazy loading (by setting `max_instances_in_memory` to some positive integer) with `num_workers > 0`, then the only workers you'll have are data loading workers, which are responsible for collecting instances from the reader with `DatasetReader.read()`, and then batching and tensorizing those instances, sending the tensor dict back to the main process.

So are you saying that the main process is hanging waiting on more data to become available? If so, have you tried increasing the number of workers?

In all of my use cases I've never found data loading to be the bottleneck. I usually just set `num_workers = 1` so that data loading is done in the background by a worker and doesn't hold up the main process. I've observed lazy loading to be about the same speed, and sometimes even faster than non-lazy loading.

---

As far as I can tell, the alternate approach you described is pretty much exactly how our old PyTorch-based data loader worked.

**Comment by aleSuglia** [comment]
@epwalsh thanks for clarifying. The only difference that I can see is that, when implementing a non-lazy dataset, AllenNLP has to load in memory the dataset as a list of `Instance` objects. With Vision+Language dataset where you have several visual features associated with the current scene, an `Instance` object would be quite big therefore making loading the entire dataset in memory infeasible.

Instead, the approach in PyTorch will be different: you will first load in memory raw objects (dictionary, lists, etc.) and then implement the `__get_item__` which is responsible for transforming the raw objects to tensors only when the `DataLoader` creates a batch. This is true both for `IterableDataset` and normal `Dataset`. So I was wondering, is there any way to have a sort of `LazyInstance` object that is materialised in memory only when required? So for instance, I believe it should be materialised only in the following situations:

1. When creating the vocabulary;
2. When the data loader has to create a batch.

**Comment by epwalsh** [comment]
> PyTorch will be different: you will first load in memory raw objects (dictionary, lists, etc.) and then implement the __get_item__ which is responsible for transforming the raw objects to tensors only when the DataLoader creates a batch.

How is this different from what AllenNLP does? Most fields are only turned into tensors when [`Instance.as_tensor_dict()`](https://github.com/allenai/allennlp/blob/decb875b4129ee9e6e286c9eea5f51b3f453c944/allennlp/data/instance.py#L88) is called. For images, if you didn't want the image tensors to be loaded in memory in the `Instance`, you could just create a `Field` that stores the image path and then only loads the image when [`Field.as_tensor()`](https://github.com/allenai/allennlp/blob/decb875b4129ee9e6e286c9eea5f51b3f453c944/allennlp/data/fields/field.py#L90) is called.

> is there any way to have a sort of LazyInstance object that is materialised in memory only when required?

What benefit does this have over lazy data loading (i.e. setting `max_instances_in_memory`)?

> This makes me think that their implementation is definitely benefiting from having multiple workers that in parallel load instances and I don't understand why AllenNLP is limited in this sense.

AllenNLP does certainly benefit from using more workers with **non-lazy** loading. But with **lazy** loading, the bottleneck in training is almost always in the model, not with data loading. So increasing the number of workers beyond 1 or 2 does will probably not help.

**Comment by aleSuglia** [comment]
I think this actually clarifies completely my issues and actually gives me a few interesting directions to make my code way faster. @epwalsh thanks a lot for your help again!

**Comment by epwalsh** [comment]
@aleSuglia no problem! And if you end up implementing a more efficient `Field` subclass for your use case, we'd love to add it to the library.
